File: Java/Structural/BridgePattern.java
package Structural;

// Implementor
interface Color {
    void applyColor();
}

// Concrete Implementor
class Red implements Color {
    @Override
    public void applyColor() {
        System.out.println("Applying Red color");
    }
}

// Abstraction
abstract class Shape {
    protected Color color;
    abstract void draw();

    public Shape(Color color) {
        this.color = color;
    }
}

// Refined Abstraction
class Circle extends Shape {
    public Circle(Color color) {
        super(color);
    }
    
    @Override
    void draw() {
        System.out.print("Drawing Circle with ");
        color.applyColor();
    }
}

class Square extends Shape {
    public Square(Color color) {
        super(color);
    }
    
    @Override
    void draw() {
        System.out.print("Drawing Square with ");
        color.applyColor();
    }
}

File: Python/Structural/BridgePattern.py
# Implementor
class Color:
    def apply_color(self):
        pass

class Red(Color):
    def apply_color(self):
        print("Applying Red color")

class Blue(Color):
    def apply_color(self):
        print("Applying Blue color")

# Abstraction
class Shape:
    def __init__(self, color):
        self.color = color
    
    def draw(self):
        pass

# Refined Abstraction
class Circle(Shape):
    def draw(self):
        print("Drawing Circle with ", end="")
        self.color.apply_color()

class Square(Shape):
    def draw(self):
        print("Drawing Square with ", end="")
        self.color.apply_color()

# Client usage
red = Red()
blue = Blue()

circle = Circle(red)
square = Square(blue)

circle.draw()
square.draw()
